python/lib/dependabot/python/file_parser.rb

```diff
@@ -91,41 +91,109 @@ def package_manager
 
       sig { returns(Ecosystem::VersionManager) }
       def detected_package_manager
-        return PeotryPackageManager.new(detect_poetry_version) if poetry_lock && detect_poetry_version
+        setup_python_environment if Dependabot::Experiments.enabled?(:enable_file_parser_python_local)
+
+        return PeotryPackageManager.new(detect_poetry_version) if potery_files && detect_poetry_version
+
+        return PipCompilePackageManager.new(detect_pipcompile_version) if pip_compile_files && detect_pipcompile_version
+
+        return PipenvPackageManager.new(detect_pipenv_version) if pipenv_files && detect_pipenv_version
 
         PipPackageManager.new(detect_pip_version)
       end
 
+      sig { returns(T.any(String, T.untyped)) }
       def detect_poetry_version
-        if poetry_lock
-          version = SharedHelpers.run_shell_command("pyenv exec poetry --version")
-                                 .to_s.split("version ").last&.split(")")&.first
+        if potery_files
+          package_manager = PeotryPackageManager::NAME
 
-          log_if_version_malformed(PeotryPackageManager.name, version)
+          version = package_manager_version(package_manager)
+                    .to_s.split("version ").last&.split(")")&.first
+
+          log_if_version_malformed(package_manager, version)
 
           # makes sure we have correct version format returned
           version if version&.match?(/^\d+(?:\.\d+)*$/)
+        end
+      rescue StandardError
+        nil
+      end
+
+      sig { returns(T.any(String, T.untyped)) }
```

Comment:
why return `untyped`?

Reply:
@abdulapopoola , This is for fail safe ,  As we are extracting package manager version from shell command, in case the output changes or there is any issue in version extraction (very low possibility) we can make sure metric collection exception does not cause update to fail. We are checking for untyped at "return PoetryPackageManager.new(detect_poetry_version) if poetry_files && **detect_poetry_version**"